include/aparse/advance_lexer.hpp

```diff
@@ -0,0 +1,97 @@
+// Copyright: 2015 Mohit Saini
+// Author: Mohit Saini (mohitsaini1196@gmail.com)
+
+#ifndef _APARSE_ADVANCE_LEXER_HPP_
+#define _APARSE_ADVANCE_LEXER_HPP_
+
+
+#include "aparse/common_headers.hpp"
+#include "aparse/regex.hpp"
+#include "aparse/error.hpp"
+#include "aparse/internal_lexer.hpp"
+
+namespace aparse {
+namespace helpers {
+
+bool ParseCharRegexFunc(const string& input, Regex* output, Error* error);
+
+}  // namespace helpers
+
+
+class AdvanceLexerScopeBase: public InternalLexerScopeBase {};
+
+template<typename LexerScope = AdvanceLexerScopeBase>
+class AdvanceLexerRules: public InternalLexerRules<LexerScope> {};
+
+template<typename LexerScope = AdvanceLexerScopeBase>
+class AdvanceLexer;
+
+template<typename LexerScope = AdvanceLexerScopeBase>
+class AdvanceLexerInstance: public InternalLexerInstance<LexerScope> {
+ public:
+  AdvanceLexerInstance() = default;
+  AdvanceLexerInstance(const AdvanceLexer<LexerScope>& lexer);
+  void Init(const AdvanceLexer<LexerScope>& lexer);
+};
+
+
+template<typename LexerScope>
+class AdvanceLexer: public InternalLexer<LexerScope> {
+ public:
+  using Instance = AdvanceLexerInstance<LexerScope>;
+  using LexerRules = AdvanceLexerRules<LexerScope>;
+  using Rule = typename LexerRules::Rule;
+  Instance CreateInstance() const {
+    return AdvanceLexerInstance<LexerScope>(*this);
+  }
+  bool Build(const LexerRules&);
+  ~AdvanceLexer() = default;
+};
+
```

Comment:
nit: Extra empty lines. It might be good to self-review the code for all such misses.

Reply:
Done. I have walked through entire code base to check such misses. I might miss some, but that is not so important.